template_feed/Microsoft.DotNet.Common.ProjectTemplates.8.0/content/ConsoleApplication-CSharp/Company.ConsoleApplication1.csproj

```diff
@@ -8,6 +8,13 @@
     <LangVersion Condition="'$(langVersion)' != ''">$(ProjectLanguageVersion)</LangVersion>
     <ImplicitUsings Condition="'$(csharpFeature_ImplicitUsings)' == 'true'">enable</ImplicitUsings>
     <Nullable Condition="'$(csharpFeature_Nullable)' == 'true'">enable</Nullable>
+    <!--#if (NativeAot) -->
+    <PublishAot>true</PublishAot>
+    <InvariantGlobalization>true</InvariantGlobalization>
+
+    <!-- Uncomment below line to make native publish outputs a lot smaller on Linux and macOS -->
+    <!-- <StripSymbols>true</StripSymbols> -->
```

Comment:
I think we should follow the same approach for ASP.NET and Hosting templates. We shouldn't have different strategies for the different templates if there are no technical reasons for it. We should decide here if this is the strategy we want to take, and apply it in all templates.

I also think that this just clutters the template. IMHO - I think `StripSymbols=true` should be the default without specifying it in the .csproj. It is what most normal .NET devs expect.

Reply:
>  I think StripSymbols=true should be the default without specifying it in the .csproj. It is what most normal .NET devs expect.

I think this is the crux of it. Lot of feedback I heard was that .NET Devs expect debuggability by default, which would mean we have to leave the symbols there. Another view is what is the typical behavior for the target platform (Linux, containers).

Reply:
Could we base it on `Release` or not?

Reply:
Setting `<StripSymbols>true</StripSymbols>` doesn't make it undebuggable thought right? The symbols will still be in a file next to the executable and discovered by the debugger. The publish target doesn't exclude the symbols file (at least in my testing) so the developer would need to manually add logic to exclude the symbols file when deploying and hence make the app undebuggable. Having the files be separate makes it clear where the app size is coming from IMO. Certainly now, during previews, is the time to try out ideas like @eerhardt's and assess feedback.

Reply:
Are .NET symbols generally larger than other native stacks, e.g. C++, Go?

Reply:
Somewhat, because we tend to have very long names. This could probably be significantly improved, but we have spent zero time on it.

Reply:
In addition to the typical name lengths, another factor is complexity of types used by your program.

Here are a few numbers for a few .NET runtime binaries that are written in C++:  
- libcoreclr.so - 7,054,328
- libcoreclr.so.dbg - 96,091,944 (symbols size is 13x binary size)
- libclrjit.so - 3,416,848
- libclrjit.so.dbg - 27,781,352 (symbol size is 8x binary size)

`dotnet new api -aot` (NET 8 P3):
- todo - 12,361,544
- todo.dbg - 37,794,328 (symbol size is 3x binary size)

You can see that we are not doing that bad compared to C++.

.NET runtime uses some C++ templates, but it is not a modern C++. I would expect that the multiplier for a typical modern C++ program is going to be even higher than the multiplier for libcoreclr.so. C++ STL types tend to be very complicated, and thus have very long names.

golang had no generics until very recently. It means you could not create complicated types in it up until very recently. The naming convention in golang tends to be very terse. Both these factors together mean that the multiplier for golang symbols is expected to be on small side, smaller than what we get.

Reply:
Some more insights:
- golang compresses dwarf symbols by default (see https://github.com/golang/go/issues/26318). It makes their relatively small symbols even smaller. Dwarf symbols - especially the ones with long names - are highly compressible. The compression ratio is 6x for `dotnet new api -aot` template. The downside is that it increases build time and slows down symbol loading in debugger.
- We have prior art with managed (portable) PDBs. They are separate files by default, but one can opt-in into embedding them via `<DebugType>embedded</DebugType>`. The managed embedded portable PDBs are compressed by default as well (it adds to the build time).

I have been mulling over this over last several days. I am leaning towards changing the SDK default to `<StripSymbols>true</StripSymbols>` (ie follow .NET / Windows convention even though it differs from Unix conventions). The prior art with portable PDBs is what made me to flip.

We should consider exposing options for symbol compression and default to compressed symbols, either all the time or at least when the symbols are not stripped.

Reply:
If we decide to default `StripSymbols` to true on the SDK side, I have some additional questions:

* .NET developers are used to symbols having the .pdb extension, no matter their underlying representation (managed/portable are very different file formats). Should we also name the native symbol file with the Windows (and by extension .NET) convention? If we don't want people to learn the platform conventions, this follows the principle of least surprise.
* Is there any point in the AOT console template? My main motivation for it was really just about the discoverability of `StripSymbols`. `<PublishAot>true</PublishAot>` doesn't seem to be interesting enough as a differentiator. I've also put `<InvariantGlobalization>true</InvariantGlobalization>` because it gets rid of ICU on Linux and helps discoverability, but that one is basically just getting a ride. Are the things that are left still important enough that we need a template?

Reply:
I think it's still worth adding the option to the template if for nothing else that visibility and discovery that native AOT is now a first-class option.

Reply:
> Should we also name the native symbol file with the Windows (and by extension .NET) convention? 

This would not work well even if we wanted to. Many tools infer the file format from the extension. If we started giving .pdb extension to DWARF files, it would be endless source of breaks.

Reply:
We could maybe go with .pdb.dbg/.pdb.dwarf. It's just whether we want people to know these are the debugging symbols, not a vital part of a non-single-file app. We assume .NET developers don't know about the underlying platform. But I guess this would be orthogonal to this PR and at this point it would also be a "breaking" change from 7.0 (so will be the new default though).

@dotnet/templating-engine-maintainers could someone have a look? I think we settled on the non-technical questions and we might tweak this based on the defaults in the coming weeks.

Reply:
> We could maybe go with .pdb.dbg/.pdb.dwarf. It's just whether we want people to know these are the debugging symbols, not a vital part of a non-single-file app.

I do not think we need to go this far. People will need to learn some amount of platform specific details in any case. The problem is that we do not have a good trigger for learning about symbols today. People do not read the manual; they just try to do stuff. We have seen many variants of a story that goes like this: Somebody builds a native AOT binary on Linux for the first time, sees a single giant binary in the publish folder, and concludes that this technology is unusable. I think that having two files - one smaller executable and one large non-executable - would be good enough to trigger the learning even without the .pdb hint. Seeing two files will make people ask why there are two files, whether they need both, etc.